### PR TITLE
Allow setting of scroll position based on browser history state.

### LINKED ISSUE #3659: scroll position/state is not saved when navigating
### Observed behavior

scroll state is not saved when browsing back/forward using vue router

As [noted](https://github.com/learningequality/kolibri/pull/3636#issuecomment-388961868) by @jonboiser:

> `savedPosition` is always `{x:0, y:0}` when using back/forward button. This might have something to do with our unique usage of vue-router. Maybe if we used `router-view`, `savedPosition` would be correct

### Expected behavior

Scroll state should be preserved.

### User-facing consequences

extra scrolling and lost context

### Steps to reproduce

* going to Learn
* scrolling down on 'recommended'
* clicking something
* clicking back

I would expect it to take me to the scrolled-down location, but it goes to the top of recommended

### Context

kolibri 0.9.0

### PR DESCRIPTION
**Pull request by rtibbles**
### Summary
Monitors CoreBase scrolling, and uses that information to update window.pageYOffset to update the history state.
Sets a vuex state element to indicate the desired starting scroll position for the core base - this might be better implemented as an event, but this was easier.
CoreBase adds a watcher to this vuex state in the created property to set the scrollTop to this value when the router has set it.

### Reviewer guidance
Does it provide the intended scrolling behaviour?
Could this be implemented in a better way? In what way might this be brittle?

scrollmedo

### References
Fixes #3659

----

### Contributor Checklist

PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [ ] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.rst

### CODE REVIEW

**Review by ralphiee22 — DISMISSED** [review]
Works great for coach pages, but not really for other pages. But hey, some working is better than none working?!

**Comment by jonboiser** [comment]
Is the scroll position supposed to be unique for each page? The current behavior seems to apply the same position to the previous page even if the position was set at an even earlier page. For example

1. On Page A I go down 200 pixels, then click a link to Page B.
1. On Page B I do not scroll, but click into Page C
1. At Page C, I go back to Page B (where I did not scroll). On Page B, I am now down 200 pixels.
1. I go back to Page A and have a Y position of 0 pixels (when it was 200 at the first step)

**Comment by rtibbles** [comment]
Hrm, good point - need to clear the vuex data when leaving a page, so that it is only set from the history state.

**Comment by rtibbles** [comment]
Or rather, not the vuex state, but the window offset value - if you navigate to pages without ever scrolling, that doesn't get reset.

File: kolibri/core/assets/src/views/CoreBase/index.vue [hunk 1]
```diff
@@ -333,16 +334,54 @@
...
+        if (this.unwatchScrollHeight) {
+          this.unwatchScrollHeight();
+        }
+        if (this.loading) {
+          // Don't set scroll position until the main content
+          // of coreBase is shown in the DOM.
+          // Create a watcher to monitor changes in loading
+          // to try to set the scrollHeight after the contents
+          // have loaded.
+          this.unwatchScrollHeight = this.$watch('loading', () => {
```
Comment (indirectlylit) [inline kolibri/core/assets/src/views/CoreBase/index.vue:361]: I wonder if `unwatchScrollHeight` should be defined on `data` and initialized to `undefined` for explicitness? Seems a little weird to just slap it onto the view model.

**Review by indirectlylit — APPROVED** [review]
added 1a27eb6b65a15b93db8479f138679acccde85758 to your PR

File: kolibri/core/assets/src/views/CoreBase/index.vue (hunk 1 above)
Comment (rtibbles) [reply kolibri/core/assets/src/views/CoreBase/index.vue:361]: Why? It doesn't need to be reactive, putting it on the view itself is exactly the right place for it, the only thing that might be better is prefixing it with an underscore, but I really don't see any benefit to having it be a watched bit of internal state.

Comment (indirectlylit) [reply kolibri/core/assets/src/views/CoreBase/index.vue:361]: Yeah agreed – I don't like the fact that it's reactive either. The thing that bothered me is to add it to the `vm` namespace without declaring it anywhere.

There's a [lively discussion here](https://github.com/vuejs/vue/issues/1988) so apparently this is something that keeps people up at night :)

Looks like the original strategy you used is likely fine, and another common pattern is to put it inside `$options`

Comment (rtibbles) [reply kolibri/core/assets/src/views/CoreBase/index.vue:361]: I do find it a little odd that the github issue is so long. Either you put it in a namespace, or you put it right on the object. Vue doesn't, and shouldn't care!

**Pull request merged**